### ISSUE TITLE
NPE thrown from TreeUtils.getAssignmentContext() during SourceChecker.typeProcess

### ISSUE THREAD

**Issue by CharlesZ-Chen** [typetools/checker-framework#1546]
Take code:

```java
class Clazz {

    <T> void m(T t) { }

    {
        try {
            new Runnable() {
                public void run() {
                }
            };
        }

        finally {
            m("Hi");
        }
    }
   
}
```

Running Checker Framework `2.2.0` with any checker would cause CF crashes:

```
error: SourceChecker.typeProcess: unexpected Throwable (NullPointerException) while processing C.java
  Compilation unit: C.java
  Exception: java.lang.NullPointerException; Stack trace: org.checkerframework.javacutil.TreeUtils.getAssignmentContext(TreeUtils.java:351)
  org.checkerframework.framework.util.typeinference.TypeArgInferenceUtil.assignedTo(TypeArgInferenceUtil.java:131)
  org.checkerframework.framework.util.typeinference.DefaultTypeArgumentInference.inferTypeArgs(DefaultTypeArgumentInference.java:129)
  org.checkerframework.framework.util.AnnotatedTypes.findTypeArguments(AnnotatedTypes.java:577)
  org.checkerframework.framework.type.AnnotatedTypeFactory.methodFromUse(AnnotatedTypeFactory.java:1984)
  org.checkerframework.framework.type.AnnotatedTypeFactory.methodFromUse(AnnotatedTypeFactory.java:1947)
  org.checkerframework.framework.type.GenericAnnotatedTypeFactory.methodFromUse(GenericAnnotatedTypeFactory.java:1469)
  org.checkerframework.framework.type.TypeFromExpressionVisitor.visitMethodInvocation(TypeFromExpressionVisitor.java:170)
  org.checkerframework.framework.type.TypeFromExpressionVisitor.visitMethodInvocation(TypeFromExpressionVisitor.java:45)
  com.sun.tools.javac.tree.JCTree$JCMethodInvocation.accept(JCTree.java:1477)
  com.sun.source.util.SimpleTreeVisitor.visit(SimpleTreeVisitor.java:54)
  org.checkerframework.framework.type.TypeFromTree.fromExpression(TypeFromTree.java:36)
  org.checkerframework.framework.type.AnnotatedTypeFactory.fromExpression(AnnotatedTypeFactory.java:1202)
  org.checkerframework.framework.type.AnnotatedTypeFactory.getAnnotatedType(AnnotatedTypeFactory.java:976)
  org.checkerframework.framework.flow.CFAbstractTransfer.getValueFromFactory(CFAbstractTransfer.java:202)
  org.checkerframework.framework.flow.CFAbstractTransfer.visitMethodInvocation(CFAbstractTransfer.java:888)
  org.checkerframework.framework.flow.CFAbstractTransfer.visitMethodInvocation(CFAbstractTransfer.java:94)
  org.checkerframework.dataflow.cfg.node.MethodInvocationNode.accept(MethodInvocationNode.java:79)
  org.checkerframework.dataflow.analysis.Analysis.callTransferFunction(Analysis.java:405)
  org.checkerframework.dataflow.analysis.Analysis.performAnalysis(Analysis.java:226)
  org.checkerframework.framework.type.GenericAnnotatedTypeFactory.analyze(GenericAnnotatedTypeFactory.java:1155)
  org.checkerframework.framework.type.GenericAnnotatedTypeFactory.analyze(GenericAnnotatedTypeFactory.java:1112)
  org.checkerframework.framework.type.GenericAnnotatedTypeFactory.performFlowAnalysis(GenericAnnotatedTypeFactory.java:1022)
  org.checkerframework.framework.type.GenericAnnotatedTypeFactory.checkAndPerformFlowAnalysis(GenericAnnotatedTypeFactory.java:1418)
  org.checkerframework.framework.type.GenericAnnotatedTypeFactory.preProcessClassTree(GenericAnnotatedTypeFactory.java:248)
  org.checkerframework.common.basetype.BaseTypeVisitor.visitClass(BaseTypeVisitor.java:285)
  org.checkerframework.common.basetype.BaseTypeVisitor.visitClass(BaseTypeVisitor.java:164)
  com.sun.tools.javac.tree.JCTree$JCClassDecl.accept(JCTree.java:720)
  com.sun.source.util.TreePathScanner.scan(TreePathScanner.java:50)
  org.checkerframework.framework.source.SourceVisitor.visit(SourceVisitor.java:66)
  org.checkerframework.framework.source.SourceChecker.typeProcess(SourceChecker.java:962)
  org.checkerframework.common.basetype.BaseTypeChecker.typeProcess(BaseTypeChecker.java:502)
  org.checkerframework.javacutil.AbstractTypeProcessor$AttributionTaskListener.finished(AbstractTypeProcessor.java:185)
  com.sun.tools.javac.api.ClientCodeWrapper$WrappedTaskListener.finished(ClientCodeWrapper.java:681)
  com.sun.tools.javac.api.MultiTaskListener.finished(MultiTaskListener.java:111)
  com.sun.tools.javac.main.JavaCompiler.flow(JavaCompiler.java:1343)
  com.sun.tools.javac.main.JavaCompiler.flow(JavaCompiler.java:1297)
  com.sun.tools.javac.main.JavaCompiler.compile2(JavaCompiler.java:902)
  com.sun.tools.javac.main.JavaCompiler.compile(JavaCompiler.java:861)
  com.sun.tools.javac.main.Main.compile(Main.java:523)
  com.sun.tools.javac.main.Main.compile(Main.java:381)
  com.sun.tools.javac.main.Main.compile(Main.java:370)
  com.sun.tools.javac.main.Main.compile(Main.java:361)
  com.sun.tools.javac.Main.compile(Main.java:56)
  com.sun.tools.javac.Main.main(Main.java:42)
1 error

```

CF `2.1.14` works fine with above code, while CF `2.2.0` crashes.

**Comment by trask** [comment]
+1, this is preventing me from updating to 2.2.0 also (same stack trace)

**Comment by smillst** [comment]
I've added a work around for this particular symptom of the null TreePath problem.  @bitterfox 
 Please revert  https://github.com/typetools/checker-framework/commit/157049f8444dd85562da761be300d79685368a9c when fixing this issue.

**Comment by smillst** [comment]
Closed in favor of #1520.
